constraints/miden-vm/decoder.air

```diff
@@ -0,0 +1,534 @@
+    let u_respan = f_respan(op_bits, extra) * (alpha[0] + alpha[1] * addr + alpha[2] * hasher[1]')
+    let v_respan = f_respan(op_bits, extra) * (alpha[0] + alpha[1] * addr' + alpha[2] * hasher[1]')
+
+    # When END operation is executed, row (addr, addr', hasher[5]) is removed from the block span 
+    # table. Register hasher[5] contains the is_loop flag.
+    let u_end = f_end(op_bits, extra) * 
+        (alpha[0] + alpha[1] * addr + alpha[2] * addr' + alpha[3] * hasher[5])
+
+    # Enforce the block stack table constraint.
+    # TODO: do proper line wrapping
+    enf p[1]' * (u_end + u_respan + 1 - (f_end(op_bits, extra) + f_respan(op_bits, extra)))
+        = p[1] * (v_join + v_split + v_loop + v_span + v_respan + 1 - (f_join(op_bits) + 
+        f_split(op_bits) + f_loop(op_bits) + f_span(op_bits) + 
+        f_respan(op_bits, extra)))
+
+
+# Enforces constraint for updating the block hash table.
+#
+# Constraint degree: 9
+ev block_hash_table_constraints([addr, op_bits[7], hasher[8], s0], [extra, p[4]]):
+    let op_bits_next = [op_bits[0]', op_bits[1]', op_bits[2]', op_bits[3]', op_bits[4]', op_bits[5]', op_bits[6]']
+    let extra_next = extra'
+
+    # Values representing left and right children of a block.
+    let ch1 = alpha[0] + alpha[1] * addr' + sum([alpha[i + 2] * hasher[i] for i in 0..4])
+    let ch2 = alpha[0] + alpha[1] * addr' + sum([alpha[i + 2] * hasher[i + 4] for i in 0..4])
+
+    # Value representing the result of hash computation.
+    let bh = alpha[0] + alpha[1] * addr + sum([alpha[i + 2] * hasher[i]]) + alpha[7] * hasher[4]
+
+    # When JOIN operation is executed, hashes of both child nodes are added to the block hash 
+    # table.
+    let v_join = f_join(op_bits) * (ch1 + alpha[6]) * ch2
+
+    # When SPLIT operation is executed and the top of the stack is 1, hash of the true branch is 
+    # added to the block hash table, but when the top of the stack is 0, hash of the false branch
+    # is added to the block hash table.
+    let v_split = f_split(op_bits) * (s0 * ch1 + (1 - s0) * ch2)
+
+    # When LOOP operation is executed and the top of the stack is 1, hash of loop body is added to
+    # the block hash table.
+    let v_loop = f_loop(op_bits) * s0 * (ch1 + alpha[7])
+
+    # When REPEAT operation is executed, hash of loop body is added to the block hash table.
+    let v_repeat = f_repeat(op_bits, extra) * (ch1 + alpha[7])
+
+    # When END operation is executed, hash of the completed block is removed from the block hash 
+    # table.
+    let u_end = f_end(op_bits, extra) * 
+        (bh + alpha[6] * (1 - (f_end(op_bits_next, extra_next) + f_repeat(op_bits_next, extra_next))))
+
+    # Enforce the block hash table constraint.
+    enf p[2]' * (u_end + 1 - f_end(op_bits, extra)) = 
+        p[2] * (v_join + v_split + v_loop + v_repeat + 1 - 
+        (f_join(op_bits) + f_split(op_bits) + f_loop(op_bits) + f_repeat(op_bits, extra)))
```

Comment:
same comment here about saving the flag computations as variables above instead of recomputing them